### PR TITLE
Extend weapon functionality

### PR DESCRIPTION
**Pull request by aske02**
This implements some new features for the weapon extension.

It has some features mentioned in #2677 and some extra ones.

**New Functions**
```js
entity:hasWeapon(string classname)

entity:clip1Size()

entity:setClip1(number amount)

entity:clip2Size()

entity:setClip2(number amount)

entity:giveWeapon(string classname)

entity:giveWeapon(string classname, number noAmmo)

entity:selectWeapon(string classname)

entity:giveAmmo(number amount, string type)

entity:giveAmmo(number amount, string type, number hidePopUp)

entity:setAmmo(number ammoCount, string type)

entity:removeAmmo(number ammoCount, string type)

entity:removeAllAmmo()

entity:stripWeapon(string classname)

entity:stripWeapons()
```

**New Events**
``` js
event weaponPickup(Weapon:entity, Owner:entity) {
    
}

event weaponSwitched(Player:entity, OldWeapon:entity, NewWeapon:entity) {
    
}
```

**New Convars**
`wire_expression2_weapon_ammo_set_enable`
`wire_expression2_weapon_ammo_give_enable`
`wire_expression2_weapon_give_enable`

### CODE REVIEW

**Comment by Vurv78** [comment]
Feel like `weaponEquipped` should be `weaponPickup` instead, to not be confused with `weaponSwitched`.

As for permissions:

There definitely needs to be a convar for these functions, probably a separate one for each type of function:
1. Ammo giving/setting, removing shouldn't need anything.
2. Giving weapon

Also need to check if you `WireLib.CanTool` or `E2Lib.isFriend` with the target player for all of these functions.

For giving weapon also need to check https://wiki.facepunch.com/gmod/SANDBOX:PlayerGiveSWEP

**Comment by aske02** [comment]
My bad hadn't really thought about permissions.
They should be added now.

File: lua/entities/gmod_wire_expression2/core/weapon.lua [hunk 1]
```diff
@@ -78,3 +134,114 @@ e2function string entity:tool()
...
+	if not giveAmmoCVar:GetBool() then return self:throw("The server has disabled giving ammo", 0) end
+	if not IsValid(this) then return self:throw("Invalid entity!", nil) end
+	if not this:IsPlayer() then return self:throw("Expected a Player but got Entity", nil) end
+	if not CanTool(self.entity.player, this, "wire_expression2") then return self:throw("You cannot target this player", nil) end
+
+	this:SetAmmo(ammoCount, type)
+end
+
+e2function void entity:removeAmmo(ammoCount, string type)
+	if not IsValid(this) then return self:throw("Invalid entity!", nil) end
```
Comment (Vurv78) [inline lua/entities/gmod_wire_expression2/core/weapon.lua:199]: Second thought, probably do want to check `setAmmoCvar` on the remove* functions. Could imagine it going quite bad when a server gives people CPPI access and some guy just forces everyones ammo to 0.

File: lua/entities/gmod_wire_expression2/core/weapon.lua [hunk 2]
```diff
@@ -78,3 +134,114 @@ e2function string entity:tool()
...
+e2function void entity:removeAllAmmo()
+	if not IsValid(this) then return self:throw("Invalid entity!", nil) end
+	if not this:IsPlayer() then return self:throw("Expected a Player but got Entity", nil) end
+	if not CanTool(self.entity.player, this, "wire_expression2") then return self:throw("You cannot target this player", nil) end
+
+	this:RemoveAllAmmo()
+end
+
+e2function void entity:stripWeapon(string classname)
+	if not IsValid(this) then return self:throw("Invalid entity!", nil) end
```
Comment (Vurv78) [inline lua/entities/gmod_wire_expression2/core/weapon.lua:215]: Same as the removeAmmo comment, want a convar check here

**Review by Vurv78 — CHANGES_REQUESTED** [review]
Just need to fix the lints, check E2Lib.isFriend before CanTool and add convars to the functions I previously thought were fine to leave without any convars.

**Pull request merged**